qutip/propagator.py

```diff
@@ -196,22 +196,22 @@ def propagator(H, t, c_op_list=[], args={}, options=None,
         sqrt_N = int(np.sqrt(N))
         dims = H0.dims
 
-        u = np.zeros([N, N, len(tlist)], dtype=complex)
-
         if parallel:
-            output = parallel_map(_parallel_mesolve, range(N * N),
+            output = parallel_map(_parallel_mesolve, range(N),
                                   task_args=(
                                       sqrt_N, H, tlist, c_op_list, args,
                                       options),
+                                  task_kwargs={"dims": H0.dims[0]},
                                   progress_bar=progress_bar, num_cpus=num_cpus)
-            for n in range(N * N):
+
+            u = np.zeros([N, N, len(tlist)], dtype=complex)
+            for n in range(N):
                 for k, t in enumerate(tlist):
                     u[:, n, k] = mat2vec(output[n].states[k].full()).T
         else:
-            rho0 = qeye(N, N)
-            rho0.dims = [[sqrt_N, sqrt_N], [sqrt_N, sqrt_N]]
-            output = mesolve(H, psi0, tlist, [], args, options,
-                             _safe_mode=False)
+            rho0 = qeye(H0.dims[0])
+            output = mesolve(H, rho0, tlist, None, [], args, options,
+                             _safe_mode=True)
```

Comment:
For reading clarity, it might be better to remove the `c_ops` and `e_ops` arguments here, and pass `args` and `options` by keyword.  (I personally think `c_ops` and `e_ops` should be keyword-only arguments to `mesolve`, but I'm not going to force that on anyone.)

In theory you could keep `_safe_mode=False` here (`propagator` has checked the dimensions of `H` and `rho`, so `mesolve` doesn't need to), which is consistent with the rest of the file.  That said, the cost of the checks is very low, but we'll hopefully get rid of it in `dev.major`; a stateful object encapsulating `mesolve` should allow users to skip these checks, in addition to us.

Reply:
Thanks for picking up the `_safe_mode` change -- I initially changed it only for debugging purposes. Restored to `False`.

Agreed about passing `args` and `options` by keyword -- the long lists of unnamed arguments was a bit terrifying, so I've implemented that in this PR too.